### PR TITLE
Adding a capability to add external store to informers

### LINKED ISSUE #114659: Unable to Set External Store in client-go
### What would you like to be added?

In shared_informer.go, when the resources are fetched from API Server, they are always cached in Memory and there is no mechanism to control that store. 

```
func NewSharedIndexInformer(lw ListerWatcher, exampleObject runtime.Object, defaultEventHandlerResyncPeriod time.Duration, indexers Indexers) SharedIndexInformer {
	realClock := &clock.RealClock{}
	sharedIndexInformer := &sharedIndexInformer{
		processor:                       &sharedProcessor{clock: realClock},
		indexer:                         NewIndexer(DeletionHandlingMetaNamespaceKeyFunc, indexers),
		listerWatcher:                   lw,
		objectType:                      exampleObject,
		resyncCheckPeriod:               defaultEventHandlerResyncPeriod,
		defaultEventHandlerResyncPeriod: defaultEventHandlerResyncPeriod,
		cacheMutationDetector:           NewCacheMutationDetector(fmt.Sprintf("%T", exampleObject)),
		clock:                           realClock,
	}
	return sharedIndexInformer
}
```
It is better to have a provision to supply the store as well something like below.
I would like to write the resource information to disk instead of memory to reduce the memory usage.

```
func (s *sharedIndexInformer) SetStore(store ThreadSafeStore) error {
	s.startedLock.Lock()
	defer s.startedLock.Unlock()

	if s.started {
		return fmt.Errorf("informer has already started")
	}
	s.indexer = NewIndexerWithStore(DeletionHandlingMetaNamespaceKeyFunc, store)
	return nil
}
```

### Why is this needed?

If we use a client-go informer to query Kubernetes resources, the memory utilization is very high for larger clusters.
This would make informers unusable for watching Kubernetes resources.

### PR DESCRIPTION
**Pull request by jeevanragula**
#### What type of PR is this?
kind/feature

#### Which issue(s) this PR fixes:

Fixes #114659

#### Does this PR introduce a user-facing change?

```release-note
We can set external store when creating shared informers now to reduce the memory utilization. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by cici37** [comment]
/assign @MikeSpreitzer
Would you mind to take a look when have time? Thanks
/triage accepted

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go [hunk 1]
```diff
@@ -455,6 +455,17 @@ func (s *sharedIndexInformer) SetTransform(handler TransformFunc) error {
 	return nil
 }
 
+func (s *sharedIndexInformer) SetStore(store ThreadSafeStore) error {
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/client-go/tools/cache/shared_informer.go:458]: I think it is better to construct the right thing from the start than to side-effect an informer after it is constructed.  I recommend extending SharedIndexInformerOptions with the option to specify the Store.

**Review by MikeSpreitzer — COMMENTED** [review]
I think this begs the question of what happens if the store already has some contents when the informer is constructed.

One use case that I have considered is tolerating intermittent connectivity.  With a local disk-based Store, for example, a client could start up and run even while the apiservers are unreachable.  But that would best be done by the informer delivering notifications of the objects that are already in the Store.  But the informer should probably try the apiservers for a little while before giving up and delivering indeterminately stale notifications.

What are you thinking a disk-based Store would do if the relevant branch of the filesystem is or goes offline?  And what if it comes back later?

**Comment by dims** [comment]
Please sign the CLA and then reopen this PR. thanks!

**Pull request closed without merging**